sql/opt: propagate row-level locking mode to index, lookup, inverted, and zigzag joins

nvb opened the pull request:
Fixes #56941.

The first commit updates the execbuilder to propagate row-level locking modes
through transformations from standard Scan and Join operations to
specialized IndexJoin, LookupJoin, and InvertedJoin operations.

The second commit updates the execbuilder to propagate row-level locking modes
through transformations from standard Scan and Join operations to
ZigZagJoins, and allows for the use of zigzag joins when explicit
row-level locking modes are in use.

Release note (sql change): table scans performed as a part of index
joins, lookup joins, and inverted joins now respect the row-level
locking strength and wait policy specified by the optional
FOR SHARE/UPDATE [NOWAIT] clause on SELECT statements.

Release note (sql change): table scans performed as a part of zigzag
joins now respect the row-level locking strength and wait policy
specified by the optional FOR SHARE/UPDATE [NOWAIT] clause on SELECT
statements.

nvb committed:
sql/opt: propagate row-level locking mode to index, lookup, and inverted joins

Fixes #56941.

This commit updates the execbuilder to propagate row-level locking modes
through transformations from standard Scan and Join operations to
specialized IndexJoin, LookupJoin, and InvertedJoin operations.

Release note (sql change): table scans performed as a part of index
joins, lookup joins, and inverted joins now respect the row-level
locking strength and wait policy specified by the optional
FOR SHARE/UPDATE [NOWAIT] clause on SELECT statements.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60719)
<!-- Reviewable:end -->

RaduBerinde reviewed (commented):
The opt part looks good, except that we should add the locking info in expr_format.go and add some xform tests.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60719#-:-MTqT8AW7Y6qZvbDUxIQ:bqd32sg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/sql/opt/ops/relational.opt, line 291 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60719#-MTqSnlf5PRHoN2e35XR:-MTqSnlf5PRHoN2e35XS:b-krujad) ([raw file](https://github.com/cockroachdb/cockroach/blob/e23c168fb5ad79147457dd8b084391f40b809086/pkg/sql/opt/ops/relational.opt#L291)):*
> ```
>     # retrieval of DELETE and UPDATE statements. The locking item's Targets list
>     # will always be empty when part of a IndexJoinPrivate.
>     Locking LockingItem
> ```

It seems that we only want the Strength and Wait policy here. I'd consider splitting those into a separate struct and using that.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
Thanks for the review Radu. Apologies for letting this sit for so long. I've revived it and addressed your two comments.

> The opt part looks good, except that we should add the locking info in expr_format.go and add some xform tests.

Done.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60719#-:-N048yol7Ksffsd-PwEQ:b-g11k4j)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde)

___
*[pkg/sql/opt/ops/relational.opt, line 291 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60719#-MTqSnlf5PRHoN2e35XR:-N04949H6hhOYK9fNpQx:bz4384k) ([raw file](https://github.com/cockroachdb/cockroach/blob/e23c168fb5ad79147457dd8b084391f40b809086/pkg/sql/opt/ops/relational.opt#L291)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It seems that we only want the Strength and Wait policy here. I'd consider splitting those into a separate struct and using that.
</blockquote></details>

Done in https://github.com/cockroachdb/cockroach/pull/79679.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
I think the only remaining operator that talks to KV that doesn't yet support locking is zigzag join - do we want to add the plumbing there too?

> Reviewed 18 of 18 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60719#-:-N06muN6CYUKSP8SZ3y2:byw4nsa)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->

nvb committed:
sql/opt: propagate row-level locking mode to zigzag joins

This commit updates the execbuilder to propagate row-level locking modes
through transformations from standard Scan and Join operations to
ZigZagJoins, and allows for the use of zigzag joins when explicit
row-level locking modes are in use.

Release note (sql change): table scans performed as a part of zigzag
joins now respect the row-level locking strength and wait policy
specified by the optional FOR SHARE/UPDATE [NOWAIT] clause on SELECT
statements.

nvanbenschoten commented:
> I think the only remaining operator that talks to KV that doesn't yet support locking is zigzag join - do we want to add the plumbing there too?

Done. I added support for SFU propagation into zigzag joins in a second commit.

yuzefovich reviewed (approved):
Thanks! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 11 of 11 files at r3, 15 of 15 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60719#-:-N0EdePa1IP7wdE50Lq6:b-ftzx47)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/sql/opt/xform/rules/select.opt, line 62 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60719#-N0EcoZLCfhcx7X9G6tC:-N0EcoZLCfhcx7X9G6tD:b72c7dn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6253d676985d5e65f1337aaebf7ad17c53e85d0/pkg/sql/opt/xform/rules/select.opt#L62)):*
> ```
> # planned.
> #
> # Zigzag joins are prohibited when the source Scan operator has been configured
> ```

nit: the comment needs an adjustment.
___
*[pkg/sql/opt/xform/rules/select.opt, line 79 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60719#-N0EdRMf2jQvpa5VOWv9:-N0EdRMf2jQvpa5VOWvA:b-5v3gaq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6253d676985d5e65f1337aaebf7ad17c53e85d0/pkg/sql/opt/xform/rules/select.opt#L79)):*
> ```
> # at any two of the JSON paths we are querying for.
> #
> # Zigzag joins are prohibited when the source Scan operator has been configured
> ```

nit: ditto.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
Very nice!

> Reviewed 8 of 18 files at r2, 11 of 11 files at r3, 15 of 15 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60719#-:-N0Gl9708MMq53BG1_Qe:b-j7qk9t)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/sql/opt/ops/relational.opt, line 648 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60719#-N0GkYH4E1OB_oSsNQkD:-N0GkYH560rV3QfMYedY:b-aqnsm8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6253d676985d5e65f1337aaebf7ad17c53e85d0/pkg/sql/opt/ops/relational.opt#L648)):*
> ```
>     # conflicting locks.
>     LeftLocking Locking
>     RightLocking Locking
> ```

A zigzag join always operates on two indexes of the same table, so these locking modes should always be the same. Can we make it a single field instead of two, then?
___
*[pkg/sql/opt/optbuilder/testdata/select_for_update, line 1005 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60719#-N0GgPGk1qW1QhOqmWlE:-N0GgPGk1qW1QhOqmWlF:b-hf1t66) ([raw file](https://github.com/cockroachdb/cockroach/blob/84d64637403bcf70a17d203020c1cd2570d5835c/pkg/sql/opt/optbuilder/testdata/select_for_update#L1005)):*
> ```
> 
> # ------------------------------------------------------------------------------
> # Tests with lookup joins.
> ```

I think a better place for these would be `pkg/sql/opt/xform/testdata/rules/join` and `.../xform/testdata/rules/select`, in the sections with the associated rule tests, e.g. [GeneratePartialIndexScans](https://github.com/cockroachdb/cockroach/blob/2115a44e6993dd4355365b2748fbe8b9aed288ca/pkg/sql/opt/xform/testdata/rules/select#L137), [GenerateConstrainedScans](https://github.com/cockroachdb/cockroach/blob/2115a44e6993dd4355365b2748fbe8b9aed288ca/pkg/sql/opt/xform/testdata/rules/select#L518), [GenerateInvertedIndexScans](https://github.com/cockroachdb/cockroach/blob/2115a44e6993dd4355365b2748fbe8b9aed288ca/pkg/sql/opt/xform/testdata/rules/select#L2119), [GenerateLookupJoins](https://github.com/cockroachdb/cockroach/blob/2115a44e6993dd4355365b2748fbe8b9aed288ca/pkg/sql/opt/xform/testdata/rules/join#L2280), [GenerateInvertedJoins](https://github.com/cockroachdb/cockroach/blob/2115a44e6993dd4355365b2748fbe8b9aed288ca/pkg/sql/opt/xform/testdata/rules/join#L6660). Then, it'll be more nature to use the `opt` directive instead of the `build` directive so the test will actually show the join and the correct locking. Alternatively, you could make a new test file like `xform/testdata/rules/for_update_locking` for these tests if you think they're better kept together.

Also, it be good to add test cases when GenerateLookupJoins and GenerateInvertedJoins build a lookup/inverted join and an index join (because the lookup index is not covering).


<!-- Sent from Reviewable.io -->

nvb committed:
sql/opt: add xform testing for select_for_update

This commit adds a new `pkg/sql/opt/xform/testdata/rules/select_for_update` test
to `pkg/sql/opt/xform`'s TestRules test suite. This allows the select for update
tests to use the `opt` directive to show joins, assert the use of specific
rules, and assert the propagation of row-level locking modes into joins.

nvb reviewed (commented):
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60719#-:-N0HYYjyA2Jgc-FiBEsU:bz5mair)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @RaduBerinde, and @yuzefovich)

___
*[pkg/sql/opt/ops/relational.opt, line 648 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60719#-N0GkYH4E1OB_oSsNQkD:-N0HOZNJA3L5eVyakg3J:b-2p8ivx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6253d676985d5e65f1337aaebf7ad17c53e85d0/pkg/sql/opt/ops/relational.opt#L648)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

A zigzag join always operates on two indexes of the same table, so these locking modes should always be the same. Can we make it a single field instead of two, then?
</blockquote></details>

Do we want to restrict ourselves to this limited form of zigzag joins indefinitely? My reading of the rest of this struct is that we'd like to keep it as general as possible (with the caveat that it still assumes only 2 join sides) and minimize the places where we assume the tables are the same. The best example of that is the `LeftTable` and `RightTable` fields in `ZigzagJoinPrivate`, which are always identical for the current use of zigzag joins.
___
*[pkg/sql/opt/xform/rules/select.opt, line 62 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60719#-N0EcoZLCfhcx7X9G6tC:-N0HOQ6mEcEkley02QNi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6253d676985d5e65f1337aaebf7ad17c53e85d0/pkg/sql/opt/xform/rules/select.opt#L62)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: the comment needs an adjustment.
</blockquote></details>

Done.
___
*[pkg/sql/opt/xform/rules/select.opt, line 79 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60719#-N0EdRMf2jQvpa5VOWv9:-N0HOQjo1QrUP178Ohn_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6253d676985d5e65f1337aaebf7ad17c53e85d0/pkg/sql/opt/xform/rules/select.opt#L79)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: ditto.
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/testdata/select_for_update, line 1005 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60719#-N0GgPGk1qW1QhOqmWlE:-N0HVAxU0JIHsGwcyvfx:bbj2ktu) ([raw file](https://github.com/cockroachdb/cockroach/blob/84d64637403bcf70a17d203020c1cd2570d5835c/pkg/sql/opt/optbuilder/testdata/select_for_update#L1005)):*
> Alternatively, you could make a new test file like xform/testdata/rules/for_update_locking for these tests if you think they're better kept together.

Done in a separate commit. I think it's nice to keep the for_update tests together because we already have a lot of coverage for join rules and the primary thing we want to verify in all of these tests is that row-level locking properties propagate down into scans.

> Also, it be good to add test cases when GenerateLookupJoins and GenerateInvertedJoins build a lookup/inverted join and an index join (because the lookup index is not covering).

Good idea. This was already being tested in the inverted join case, but not in the lookup join case. Done.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> Reviewed 2 of 17 files at r5, 13 of 15 files at r6, 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60719)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @nvanbenschoten, and @yuzefovich)

___
*[pkg/sql/opt/ops/relational.opt, line 648 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60719#-N0GkYH4E1OB_oSsNQkD:-N0HgbAaD2eCuzgzo_jL:brhhd0w) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6253d676985d5e65f1337aaebf7ad17c53e85d0/pkg/sql/opt/ops/relational.opt#L648)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we want to restrict ourselves to this limited form of zigzag joins indefinitely? My reading of the rest of this struct is that we'd like to keep it as general as possible (with the caveat that it still assumes only 2 join sides) and minimize the places where we assume the tables are the same. The best example of that is the `LeftTable` and `RightTable` fields in `ZigzagJoinPrivate`, which are always identical for the current use of zigzag joins.
</blockquote></details>

Good point, I wasn't thinking about the possibility of multi-table zig zag joins. Let's leave as-is!
___
*[pkg/sql/opt/optbuilder/testdata/select_for_update, line 1005 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60719#-N0GgPGk1qW1QhOqmWlE:-N0HcYfD6XJij7CqUI5G:b7gbova) ([raw file](https://github.com/cockroachdb/cockroach/blob/84d64637403bcf70a17d203020c1cd2570d5835c/pkg/sql/opt/optbuilder/testdata/select_for_update#L1005)):*
> the primary thing we want to verify in all of these tests is that row-level locking properties propagate down into scans.

Ahh I misunderstood. There's two things we're testing:

1. That modes propagate down into scans during optbuilding. These belong in optbuilder tests, and is covered by these tests you originally added.  I think the titles like "Tests with lookup joins" confused me — something like "Tests with joins" might be better?
2. That lookup, index, and inverted joins inherit the locking modes of the scans they are derived from. These are the cases that I thought should be in xform tests since xform generates theses operators. It sounds like you were originally covering this in the execbuilder tests. In that case, apologies for the extra work I created for you. There's some overlap in the xform tests you added, but I think that's ok.
___
*[pkg/sql/opt/xform/testdata/rules/select_for_update, line 14 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/60719#-N0HfdqQ2WY7IweAkZGy:-N0HfdqQ2WY7IweAkZGz:b-i3ojgf) ([raw file](https://github.com/cockroachdb/cockroach/blob/884af5451ba404ca58d6bdb7d596f84ea373ec6f/pkg/sql/opt/xform/testdata/rules/select_for_update#L14)):*
> ```
> 
> # ------------------------------------------------------------------------------
> # Basic tests.
> ```

To me it'd make the most sense for this file to only include cases related to exploration — tests that ensure that lookup, index, and inverted joins have the correct locking modes. All other tests sound like they are verifying that locking modes are pushed down correctly during the optbuild phase, so I think they should go there. 


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60719)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @yuzefovich)

___
*[pkg/sql/opt/optbuilder/testdata/select_for_update, line 1005 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60719#-N0GgPGk1qW1QhOqmWlE:-N0I3Aaq89WKSVYhm-Ch:b-wmcwfv) ([raw file](https://github.com/cockroachdb/cockroach/blob/84d64637403bcf70a17d203020c1cd2570d5835c/pkg/sql/opt/optbuilder/testdata/select_for_update#L1005)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

> the primary thing we want to verify in all of these tests is that row-level locking properties propagate down into scans.

Ahh I misunderstood. There's two things we're testing:

1. That modes propagate down into scans during optbuilding. These belong in optbuilder tests, and is covered by these tests you originally added.  I think the titles like "Tests with lookup joins" confused me — something like "Tests with joins" might be better?
2. That lookup, index, and inverted joins inherit the locking modes of the scans they are derived from. These are the cases that I thought should be in xform tests since xform generates theses operators. It sounds like you were originally covering this in the execbuilder tests. In that case, apologies for the extra work I created for you. There's some overlap in the xform tests you added, but I think that's ok.
</blockquote></details>

That's well put. These xform tests are specifically interested in #2, because that can't be tested in the optbuilder tests. They were already exercised to some degree in the execbuider tests, but that was a little less direct so I'm glad you encouraged me to add this test.
___
*[pkg/sql/opt/xform/testdata/rules/select_for_update, line 14 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/60719#-N0HfdqQ2WY7IweAkZGy:-N0I3oxw-V8Y2Cwx6g5g:b-631bma) ([raw file](https://github.com/cockroachdb/cockroach/blob/884af5451ba404ca58d6bdb7d596f84ea373ec6f/pkg/sql/opt/xform/testdata/rules/select_for_update#L14)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

To me it'd make the most sense for this file to only include cases related to exploration — tests that ensure that lookup, index, and inverted joins have the correct locking modes. All other tests sound like they are verifying that locking modes are pushed down correctly during the optbuild phase, so I think they should go there. 
</blockquote></details>

Good point. I've stripped this down to just the set of tests related to exploration rules.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Looks great!

> Reviewed 2 of 17 files at r5, 2 of 15 files at r6, 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60719#-:-N0amfne2dk3xGcc4QEH:b-2t11ox)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten)

___
*[`pkg/sql/opt/optbuilder/testdata/select_for_update` line 1005 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60719#-N0GgPGk1qW1QhOqmWlE:-N0ameIG9Inngmluc7bh:b-8fx2u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/84d64637403bcf70a17d203020c1cd2570d5835c/pkg/sql/opt/optbuilder/testdata/select_for_update#L1005)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

That's well put. These xform tests are specifically interested in #2, because that can't be tested in the optbuilder tests. They were already exercised to some degree in the execbuider tests, but that was a little less direct so I'm glad you encouraged me to add this test.
</blockquote></details>

:+1: 


<!-- Sent from Reviewable.io -->

michae2 commented:
Any chance we could backport this? At least to 22.1?

nvanbenschoten commented:
@michae2 I opened up https://github.com/cockroachdb/cockroach/pull/81348 to backport this to 22.1. It does feel a little large for a backport, but I think it's relatively low risk.
